### PR TITLE
Graduate kube-scheduler NodeInclusionPolicy to beta

### PR DESCRIPTION
**Pull request by kerthcet**
Signed-off-by: kerthcet <kerthcet@gmail.com>

#### What type of PR is this?
/kind feature
/sig scheduling

#### What this PR does / why we need it:
Graduate NodeInclusionPolicy to beta, only api part.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NodeInclusionPolicy in podTopologySpread plugin is enabled by default. 
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- KEP: https://github.com/kubernetes/enhancements/issues/3094
```

### CODE REVIEW

**Comment by kerthcet** [comment]
This is the benchmark test for nodeInclusionPolicy:
```
1) 500 nodes:
NodeInclusionPolicy disabled:
BenchmarkPerfScheduling/SchedulingWithNodeInclusionPolicy/500Nodes-8     1    42853905745 ns/op
ok  	k8s.io/kubernetes/test/integration/scheduler_perf    62.579s

NodeInclusionPolicy enabled:
BenchmarkPerfScheduling/SchedulingWithNodeInclusionPolicy/500Nodes-8     1    42976119234 ns/op
ok  	k8s.io/kubernetes/test/integration/scheduler_perf    62.711s

2) 5000 nodes:
NodeInclusionPolicy disabled:
BenchmarkPerfScheduling/SchedulingWithNodeInclusionPolicy/5000Nodes-8     1    383803119491 ns/op
ok  	k8s.io/kubernetes/test/integration/scheduler_perf    555.560s

NodeInclusionPolicy enabled:
BenchmarkPerfScheduling/SchedulingWithNodeInclusionPolicy/5000Nodes-8     1    388210043641 ns/op
ok  	k8s.io/kubernetes/test/integration/scheduler_perf    560.219s
```

File: test/integration/framework/perf_utils.go [hunk 1]
```diff
@@ -119,7 +119,7 @@ func (p *IntegrationTestNodePreparer) PrepareNodes(nextNodeIndex int) error {
 func (p *IntegrationTestNodePreparer) CleanupNodes() error {
 	// TODO(#93794): make CleanupNodes only clean up the nodes created by this
 	// IntegrationTestNodePreparer to make this more intuitive.
-	nodes, err := GetReadySchedulableNodes(p.client)
```
Comment (liggitt) [inline test/integration/framework/perf_utils.go:122]: curious why this change was needed?

File: test/integration/framework/util.go [hunk 2]
```diff
@@ -62,25 +62,6 @@ func DeleteNamespaceOrDie(c clientset.Interface, ns *v1.Namespace, t *testing.T)
...
-// 2) Needs to be ready.
-// If EITHER 1 or 2 is not true, most tests will want to ignore the node entirely.
-// If there are no nodes that are both ready and schedulable, this will return an error.
-func GetReadySchedulableNodes(c clientset.Interface) (nodes *v1.NodeList, err error) {
-	nodes, err = checkWaitListSchedulableNodes(c)
-	if err != nil {
-		return nil, fmt.Errorf("listing schedulable nodes error: %s", err)
-	}
-	Filter(nodes, func(node v1.Node) bool {
-		return IsNodeSchedulable(&node) && isNodeUntainted(&node)
```
Comment (liggitt) [inline test/integration/framework/util.go:76]: lint is now complaining that isNodeUntainted is unused

**Review by liggitt — COMMENTED** [review]
API plumbing lgtm, will defer benchmark review and test change to scheduling reviewer

File: test/integration/framework/perf_utils.go (hunk 1 above)
Comment (kerthcet) [reply test/integration/framework/perf_utils.go:122]: When creating nodes with taints, all these nodes are unschedulable to normal pods, so we'll always return error after calling `Filter()`. That means no node satisfied.

`checkWaitListSchedulableNodes` will return all nodes with `spec.unschedulable` is false, the only difference here is we'll patch more nodes than before, who will not be scheduled pods. So I think it's ok.

File: test/integration/framework/util.go (hunk 2 above)
Comment (kerthcet) [reply test/integration/framework/util.go:76]: Removed.

**Comment by kerthcet** [comment]
The test `pull-kubernetes-e2e-kind`  failed mostly related to https://github.com/kubernetes/kubernetes/issues/113547

File: test/integration/framework/perf_utils.go (hunk 1 above)
Comment (Huang-Wei) [reply test/integration/framework/perf_utils.go:122]: This actually impacts all existing tests' behavior to skip nodes' health check. I'm a bit concerned on this. Can you add a param in `workloadTemplate` to make the behavior of "ignoring node health check or not" an option?

Comment (alculquicondor) [reply test/integration/framework/perf_utils.go:122]: Should we make the parameter ignore taints, instead of ignoring isNodeSchedulable?

Comment (alculquicondor) [reply test/integration/framework/perf_utils.go:122]: although... why did we even care about getting Ready nodes in the first place? I don't think we have any test cases with non-ready nodes.

File: test/integration/framework/util.go [hunk 3]
```diff
@@ -62,25 +60,6 @@ func DeleteNamespaceOrDie(c clientset.Interface, ns *v1.Namespace, t *testing.T)
...
-	})
-	if len(nodes.Items) == 0 {
-		return nil, fmt.Errorf("there are currently no ready, schedulable nodes in the cluster")
-	}
-	return nodes, nil
-}
-
 // checkWaitListSchedulableNodes is a wrapper around listing nodes supporting retries.
 func checkWaitListSchedulableNodes(c clientset.Interface) (*v1.NodeList, error) {
 	nodes, err := waitListSchedulableNodes(c)
```
Comment (alculquicondor) [inline test/integration/framework/util.go:65]: This function looks unnecessary actually

**Review by alculquicondor — COMMENTED** [review]
/approve
/hold for nits

File: test/integration/framework/perf_utils.go (hunk 1 above)
Comment (kerthcet) [reply test/integration/framework/perf_utils.go:122]: Add a param is one approach, but try to figure out the reason why we want to filter out taint nodes here. They're schedulable actually if pod tolerates the taints. 

 I think the change won't impact present tests for we have no taint-related or node-not-ready cases. But in theory, filtering out non-ready nodes when updating patches makes sense to me for the sake of performance especially we have thousands of nodes.

Another question here is when cleaning up, we only remove the `readySchedulableNodes` nodes, but we should remove all nodes created by itself. I used to submit a PR to fix this here https://github.com/kubernetes/kubernetes/pull/112237.

So in summary:
1) When calling `PrepareNodes`, I'll get ready and schedulable nodes including nodes with taints
2) When calling `CleanupNodes`, I'll remove all nodes created by itself as https://github.com/kubernetes/kubernetes/pull/112237.

WDYT?

Comment (kerthcet) [reply test/integration/framework/perf_utils.go:122]: > I think we can prepare and cleanup all nodes, given that we always create them as ready (there are no kubelets running in our tests).

I'm ok with this. Also looking forward to @Huang-Wei 's advices.

Comment (kerthcet) [reply test/integration/framework/perf_utils.go:122]: Created a separate PR here https://github.com/kubernetes/kubernetes/pull/113615 to make this a pure one, only focus on the API part.

**Comment by kerthcet** [comment]
Hold until https://github.com/kubernetes/kubernetes/pull/113615 got merged first.
/hold

**Comment by liggitt** [comment]
/approve
for API bits
